app/assets/javascripts/school_administrator_dashboard/discipline_dashboard/SchoolDisciplineDashboard.js

```diff
@@ -41,12 +45,56 @@ export default class SchoolDisciplineDashboard extends React.Component {
     this.setState({selectedChart: selection.value, selectedCategory: null});
   }
 
-  filterIncidentDates(incidentsArray) {
-    const {nowFn} = this.context;
-    const {timeRangeKey} = this.state;
-    const range = momentRange(timeRangeKey, nowFn());
-    return incidentsArray.filter((incident) => {
-      return moment.utc(incident.occurred_at).isBetween(range[0], range[1]);
+  allDisciplineIncidents() {
+    return this.memoize(['allDisciplineIncidents'], () => {
+      const {dashboardStudents} = this.props;
+      return _.flattenDeep(_.compact(dashboardStudents.map(student => this.mergeDisciplineData(student.discipline_incidents, student))));
+    });
+  }
+
+  //Associate all attributes that we want to use for incident grouping and student list. More may be added here later.
+  mergeDisciplineData(disciplineIncidentsArray, student) {
+    if (!disciplineIncidentsArray) return;
+    //student attributes
+    const grade = student.grade;
+    const classroom = student.homeroom_label;
+    const race = student.race;
+    const first_name = student.first_name;
+    const last_name = student.last_name;
+    const latest_note = student.latest_note;
+
+    return disciplineIncidentsArray.map(incident => {
+      //incident attributes derived from raw incident data
+      const exactTime = incident.has_exact_time ? moment.utc(incident.occurred_at).startOf('hour').format('h:mm a') : "Not Logged";
+      const day = moment.utc(incident.occurred_at).format("ddd");
+      const location = incident.incident_location || "Not Recorded";
+      return {...incident, first_name, last_name, latest_note, grade, classroom, race, exactTime, day, location};
+    });
+  }
+
+  filteredDisciplineIncidents(disciplineIncidents) {
+    return this.memoize(['filteredIncidents', this.state, arguments], () => {
+      const {nowFn} = this.context;
+      const {timeRangeKey, selectedIncidentCode} = this.state;
+      const range = momentRange(timeRangeKey, nowFn());
+      return disciplineIncidents.filter(incident => {
+        if (!moment.utc(incident.occurred_at).isBetween(range[0], range[1])) return false;
+        if (incident.incident_code !== selectedIncidentCode && selectedIncidentCode !== ALL) return false;
```

Comment:
nice!

Reply:
Does the filter for the selected time/date/etc belong here too?

Reply:
I don't think so. I'm using this to trim the incidents that fall outside the date range or selected incident type. The time of day, day of week, etc are different. Those get sorted into categories, and when the user selects that part of the bar chart they filter the student list, but the incidents still show up in the bar chart.